### PR TITLE
Feature/improved staggered rhmc

### LINKED ISSUE #284: Verify the staggered phase application kernel
While implemented, this kernel has never been fully tested, and may be buggy.  This functionality is required for a fully GPU resident staggered HMC.

### PR DESCRIPTION
**Pull request by maddyscientist**
This branch is fairly wide ranging, and came from optimization of the Cineca staggered RHMC benchmark:
* More robust SVD in HISQ
* Unitarity checks now include a check for Nan
* Improved performance of multi-shift solver
  * Delay shift updates until middle of subsequent dslash to improve strong scaling
  * Stop iterating on shifted systems if iterated residuals fall below precision limit
* Optimization of HISQ and gauge force kernels through kernel fusion
* Significant of gauge force routines
* Reduced compilation time using trait mappers 
* Added `momActionQuda` for computing the momentum contribution to the action
* Added `projectSU3Quda` for projecting the gauge field onto the SU(3) manifold (uses the projection from APE smearing)
* Fixed bugs in gauge phase application kernel and added support for imaginary chemical potential (closes #284)
* In the milc interface staggered solver, use the resident gauge field until it is invalidated by constructing a new set of fat/long links
* Experimental support for fully GPU resident momentum: download when momActionQuda is computed and invalidate (disabled by default since this is not supported in mainline MILC)
* LDG loads are now enabled by default when `gauge::FloatNOrder` is used
* Improved debugging: added L1 norm reduction kernel and check for Nan when copying gauge fields (when debugging is enabled)
* Minor bug fixes
* Removal of dead code

### CODE REVIEW

**Comment by mathiaswagner** [comment]
Lots of cool things. I know how the changes came a long but ... whew, it might need a moment to digest them.

**Comment by alexstrel** [comment]
Anyone is assinged to check the branch? Which milc version does one need?

**Comment by mikeaclark** [comment]
It would be great if one of you ( @alexstrel or @mathiaswagner ) would check this against the current mainline MILC.  I think it should work as is, just suboptimally.  Though I have a lot patches that are in the pipeline to make the mainline MILC.  Before I push those patches I wanted to finish some further work that will enable the gauge field to be fully resident through RHMC (as the momentum field is now with the above experimental feature).

**Comment by alexstrel** [comment]
Perfect, so I'll try current version of milc.

**Comment by mathiaswagner** [comment]
I'll check whether I can still fire up my existing MILC tests (need to check whether I need to transfer that to different machines first in order to run it).

File: lib/dslash_policy.cuh [hunk 1]
```diff
@@ -1131,212 +1012,10 @@ struct DslashFactory {
...
-  }
-#endif
-#ifdef PTHREADS
-  bool interiorLaunched = false;
-#endif
-  int completeSum = 0;
-  while (completeSum < pattern.commDimTotal) {
-    for (int i=3; i>=0; i--) {
-      if (!dslashParam.commDim[i]) continue;
+#if 0
```
Comment (mathiaswagner) [inline lib/dslash_policy.cuh]: Can we just delete the code that is disabled with the '#if 0'? If not we should have a comment here.

File: lib/unitarize_force_quda.cu [hunk 2]
```diff
@@ -646,7 +646,7 @@ static double HOST_REUNIT_SVD_ABS_ERROR;
       void preTune() { ; }
       void postTune() { cudaMemset(fails, 0, sizeof(int)); } // reset fails counter
       
-      long long flops() const { return 4*4528*gauge.Volume(); } // FIXME: add flops counter
+      long long flops() const { return 4ll*4528*gauge.Volume(); } // FIXME: add flops counter
```
Comment (mathiaswagner) [inline lib/unitarize_force_quda.cu]: is that fixed with the 411*4528*Volume or just better than before but still wrong? (Just wondering about the FIXME comment)

File: lib/svd_quda.h [hunk 3]
```diff
@@ -291,6 +291,9 @@ void getRealBidiagMatrix(const Matrix<Cmplx,3> & mat,
   typename RealTypeId<Cmplx>::Type beta;
   Cmplx w, tau, z;
 
+  u(0,0).x = -2.0;
```
Comment (mathiaswagner) [inline lib/svd_quda.h]: using -2.0 to detect special cases should come with a comment explain what is done

File: lib/reduce_quda.cu [hunk 4]
```diff
@@ -128,6 +128,36 @@ namespace quda {
   };
 
   /**
+     Return the L1 norm of x
+  */
+  __device__ double norm1_(const double2 &a) { return fabs(a.x) + fabs(a.y); }
+  __device__ float norm1_(const float2 &a) { return (double)fabs(a.x) + (double)fabs(a.y); }
```
Comment (mathiaswagner) [inline lib/reduce_quda.cu]: norm1_ returns float, so does the casting to double of the individual summands help in the precision of the reduction?

It is however consistent with the casting for norm2_

File: lib/dslash_policy.cuh (hunk 1 above)
Comment (mikeaclark) [inline lib/dslash_policy.cuh]: There is a comment ther
```
// FIXME there is no policy version of this variant                                                                                                                                                                                                                           
```
This is a variant of the dslash that does a direct write of the packing kernel to system memory as opposed to GPU memory.  It was an experiment that may be useful in the future...  I'll add more clarification.

File: lib/unitarize_force_quda.cu (hunk 2 above)
Comment (mikeaclark) [inline lib/unitarize_force_quda.cu]: Deleted comment.

File: lib/svd_quda.h (hunk 3 above)
Comment (mikeaclark) [inline lib/svd_quda.h]: Ok, I've cleaned up this up and added clarification in the comments what's going on here.

File: lib/reduce_quda.cu (hunk 4 above)
Comment (mikeaclark) [inline lib/reduce_quda.cu]: I noticed this issue in the feature/multigrid branch, and it's fixed there.  Will apply the fix here too.

**Comment by mikeaclark** [comment]
@nmrcardoso here are some benchmarks showing the speedups of the gauge fixing with the changes I made in commit 15c9058de885dfb5be91e0f60124ee4fecc6908d to the GaugeFixQuality kernels.

Performance reported by `gauge_alg_test --recon 12`

before
```
Time: 0.906449 s, Gflop/s =  357.2, GB/s =  138.1
Time: 0.862841 s, Gflop/s =  360.6, GB/s =  133.3
Time: 1.575909 s, Gflop/s =  258.4, GB/s =   78.7
Time: 1.181364 s, Gflop/s =  341.4, GB/s =  100.6
```
after
```
Time: 0.896090 s, Gflop/s =  361.3, GB/s =  139.7
Time: 0.839907 s, Gflop/s =  370.5, GB/s =  137.0
Time: 1.536007 s, Gflop/s =  265.1, GB/s =   80.7
Time: 1.122240 s, Gflop/s =  359.4, GB/s =  105.9
```

**Comment by nmrcardoso** [comment]
Thanks a lot @mikeaclark :)

**Comment by stevengottlieb** [comment]
Hi Kate,

I am concerned reading about fixing bugs in gauge phase application
kernel.  That sounds really scary.  Would it have affected spectrum
running with MILC?

Thanks,
Steve

**Comment by mikeaclark** [comment]
Steve, this kernel hasn't never been called by MILC, so there's zero risk of it affecting any results.  While this kernel has been in QUDA for nearly 2 years it was never tested and **never been used in production** and it was just a quick add I made while in the middle of an RFP but didn't have a chance to debug it.

In the normal MILC workflow the staggered phase has always been added / removed to the gauge field  by MILC and then this is passed to QUDA.  However, it is clearly needed for achieving a fully GPU resident pipeline in HMC.  This is something I got working for another RFP in August (and added support for imaginary chemical phase at the same time).  There is still a little more work to do, but I've basically got all the building blocks now for the full resident pipeline of the gauge field (the other thing that was needed was SU(3) projection to replace MILC's reunitarization routine).

Reassured?  :)

**Comment by mathiaswagner** [comment]
I tried to run a RHMC test using that branch and ran into a segfault. Just to make sure I updated my MILC and QUDA build and will try again.

**Comment by mathiaswagner** [comment]
With the fresh build I don't get a segfault but 

```
[...]
Unitarity problem on node 0, site 4944, dir 0 tolerance=1.000000e-04
SU3 matrix:
0.800393 0.069412 0.372997 -0.026848 -0.233114 -0.400454 
0.378685 -0.251795 -0.643163 -0.089125 0.571365 -0.212456 
0.093173 0.372919 -0.661173 -0.037440 -0.643085 0.011892 
repeat in hex:
8a890de7 b7572867 8771f21e eadea425 68f24980 e513a761 
3de69183 e9406bb3 1fc6647a 96dcdc26 2e7da42f a72d648e 
1e60c4c2 05ee47ea 7da06791 c01139bf 704f4c59 9dac0ab4 
  
 
Unitarity problem on node 0, site 4944, dir 1 tolerance=1.000000e-04
SU3 matrix:
-0.000000 0.000000 0.000000 0.000000 0.000000 -0.000000 
0.000000 -0.000000 0.000000 -0.000000 -0.000000 -0.000000 
0.000000 -0.000000 0.000000 -0.000000 0.000000 -0.000000 
repeat in hex:
00000000 00000000 00000000 00000000 00000000 00000000 
00000000 00000000 00000000 00000000 00000000 00000000 
00000000 00000000 00000000 00000000 00000000 00000000 
[...]
```

:disappointed:

**Comment by mikeaclark** [comment]
@mathiaswagner Can you rerun with `HOST_DEBUG` enabled?  This adds more robust checking like checking for the presence of Nans in the unitarization and when ever the gauge field is copied from CPU <-> GPU.

**Comment by mathiaswagner** [comment]
Yes, I will do that.

**Comment by mathiaswagner** [comment]
From a run with 'HOST_DEBUG'

'''
QUDA_MILC_INTERFACE: qudaMultishiftInvert (return) 
CONGRAD5: time = 9.189820e-01 (multicg_offset_QUDA D) masses = 7 iters = 87 mflops = 4.114608e+04
 OK converged final_rsq= 2.3e-15 (cf 2.5e-15) rel = 8.1e-14 (cf 0) restarts = 0 iters= 87
QUDA_MILC_INTERFACE: qudaComputeOprod (called) 
QUDA_MILC_INTERFACE: qudaComputeOprod (return) 
QUDA_MILC_INTERFACE: qudaComputeOprod (called) 
QUDA_MILC_INTERFACE: qudaComputeOprod (return) 
QUDA_MILC_INTERFACE: qudaHisqForce (called) 
QUDA_MILC_INTERFACE: qudaHisqForce (return) 
FFTIME:  time = 1.445482e+01 (HISQ D) terms = 33 flops/site = 0 mflops = 0.000000e+00
QUDA_MILC_INTERFACE: qudaUpdateU (called) 
ERROR: Nan detected at parity=0, dir=0, x=8459, i=0 (rank 1, host nid00842, copy_gauge_inc.cu:61 in checkNan())
       last kernel called was (name=N4quda16UpdateGaugeFieldI7double2NS_11FloatNOrderIdLi18ELi2ELi18EEENS2_IdLi18ELi2ELi11EEELi8ELb0ELb0EEE
,volume=24x24x48x12,aux=threads=331776,prec=8,stride=165888)
ERROR: Nan detected at parity=0, dir=1, x=8160, i=1 (rank 0, host nid00841, copy_gauge_inc.cu:61 in checkNan())
       last kernel called was (name=N4quda16UpdateGaugeFieldI7double2NS_11FloatNOrderIdLi18ELi2ELi18EEENS2_IdLi18ELi2ELi11EEELi8ELb0ELb0EEE
,volume=24x24x48x12,aux=threads=331776,prec=8,stride=165888)
Application 5362924 exit codes: 134
Application 5362924 exit signals: Killed
'''

**Comment by mikeaclark** [comment]
Mathias:

Can you run with one GPU to see if this is a multi GPU issue?

Can you make the Fermi fix as a patch to develop and see if this issue is only in this branch?

**Comment by mikeaclark** [comment]
The other thing I meant to asK: does this problem immediately (first call to qudaUpdateU) or after sometime?

**Comment by mathiaswagner** [comment]
@mikeaclark : My run was on K20 so no Fermi patch necessary. I can run with develop to check whether this is a regression introduce in improved-staggered-rhmc. Not sure whether I can run on a single GPU - I need to check the volume / memory.

**Comment by mathiaswagner** [comment]
The issue comes up in the 8th call to 'qudaUpdateU'.

I ran with the develop branch and that completed without errors (I should check whether the results are correct).
For the single GPU run: I am afraid 48^3x12 won't run on a single K20 in double precision.

**Comment by mathiaswagner** [comment]
Like a bloody rookie I tripped over an old tunecache ... I retuned a now the tuning run and a subsequent run using the new tunecache completed.

**Comment by mathiaswagner** [comment]
While the run completes and `delta S` is reasonable I do see a performance regression:

**Develop**:
```
               initQuda Total time = 3.7373 secs
          loadGaugeQuda Total time = 48.9687 secs
             invertQuda Total time = 214.082 secs
   invertMultiShiftQuda Total time = 263.366 secs
      computeKSLinkQuda Total time = 14.9852 secs
   computeGaugeForceQuda Total time = 8.33947 secs
   updateGaugeFieldQuda Total time = 14.2041 secs
   computeStaggeredOprodQuda Total time = 11.2918 secs
   computeHISQForceQuda Total time = 44.8452 secs
                endQuda Total time = 0.325714 secs
       initQuda-endQuda Total time = 890.945 secs
                   QUDA Total time = 624.145 secs
```

**improved-staggered-rhmc**
```
               initQuda Total time = 0.953773 secs
          loadGaugeQuda Total time = 9.07024 secs
             invertQuda Total time = 222.436 secs
   invertMultiShiftQuda Total time = 323.39 secs
      computeKSLinkQuda Total time = 25.022 secs
   computeGaugeForceQuda Total time = 14.8119 secs
   updateGaugeFieldQuda Total time = 24.7242 secs
   computeStaggeredOprodQuda Total time = 18.9855 secs
   computeHISQForceQuda Total time = 51.1936 secs
                endQuda Total time = 0.318983 secs
       initQuda-endQuda Total time = 957.145 secs
                   QUDA Total time = 690.906 secs
```

Not sure what is going on here but performance seems worse for all QUDA components. I was using CUDA 6.5 on Cray XK7  with K20 (Big Red II at Indiana University).

**Comment by mathiaswagner** [comment]
retest this please

**Comment by mathiaswagner** [comment]
The above comment is for Jenkins - I might have to update the key phrases.

**Comment by mathiaswagner** [comment]
I am not sure how we want to proceed here:
* We see performance regressions so I am reluctant to merge that in.
* On the other hand we have some necessary fixes (compile for Fermi) in here which we should get into develop asap and it looks like cherry picking them might be a mess (if possible at all).

**Comment by mikeaclark** [comment]
Can you post the full timing break downs and send me the tunecache files?

**Comment by mikeaclark** [comment]
Can I just get confirmation: the Nan issue then was completely down to the stale tunecache being used?

**Comment by mathiaswagner** [comment]
Yes. It looks like that.

**Comment by mikeaclark** [comment]
Also, just checking: you did recompile the branch without `HOST_DEBUG` being enabled?

**Comment by mathiaswagner** [comment]
Updated timings with host debug ...

**improved-staggered-rhmc**
```
               initQuda Total time = 0.94276 secs
          loadGaugeQuda Total time = 5.48502 secs
             invertQuda Total time = 213.581 secs
   invertMultiShiftQuda Total time = 257.609 secs
      computeKSLinkQuda Total time = 14.418 secs
   computeGaugeForceQuda Total time = 10.4955 secs
   updateGaugeFieldQuda Total time = 14.0914 secs
   computeStaggeredOprodQuda Total time = 11.1957 secs
   computeHISQForceQuda Total time = 43.5805 secs
                endQuda Total time = 0.262154 secs
       initQuda-endQuda Total time = 854.851 secs
                   QUDA Total time = 571.66 secs
```

**develop**
```
               initQuda Total time = 1.01184 secs
          loadGaugeQuda Total time = 51.231 secs
             invertQuda Total time = 206.192 secs
   invertMultiShiftQuda Total time = 249.673 secs
      computeKSLinkQuda Total time = 14.3976 secs
   computeGaugeForceQuda Total time = 7.96804 secs
   updateGaugeFieldQuda Total time = 14.0837 secs
   computeStaggeredOprodQuda Total time = 11.204 secs
   computeHISQForceQuda Total time = 43.5323 secs
                endQuda Total time = 0.27295 secs
       initQuda-endQuda Total time = 884.652 secs
                   QUDA Total time = 599.567 secs
```

So, **there was no regression** in the code - just in my brain.

For completeness:
```
test_develop_nodebug2.out:ACCEPT: delta S = 3.349597e-01
test_improved_nodebug2.out:ACCEPT: delta S = 3.349543e-01
```

I still need to check a CPU reference for that to make sure everything is fine.

**Comment by mikeaclark** [comment]
Thanks for confirming your mental regression :smile: 

The only things that's surprising here is that the gauge force has a slight regression.  This, despite the fact that all tests I've done show that performance is better after kernel fusion (tested on GK110 and GM200).  Can you post the break downs of the gauge force?

I believe the solver regressions are because this branch switches on the fused ghost kernel for the multi-GPU dslash: this only helps when strong scaling else it can reduce performance.  I think I'll revert that change, though maybe now is the time to expose this option in the interface.

**Comment by mikeaclark** [comment]
Just to note that I'm having problems with this branch with Carleton's 2+1+1 run, but I currently believe this is probably the charm getting messed up.  The above results are 2+1 aren't they?

**Comment by mathiaswagner** [comment]
**improved**

```
   computeGaugeForceQuda Total time = 10.4955 secs
              download     = 5.318957 secs (  50.7%), with       24 calls at 2.216232e+05 us per call
                upload     = 3.084770 secs (  29.4%), with       12 calls at 2.570642e+05 us per call
                  init     = 0.015169 secs ( 0.145%), with       36 calls at 4.213611e+02 us per call
               compute     = 1.446999 secs (  13.8%), with       12 calls at 1.205832e+05 us per call
                  free     = 0.009055 secs (0.0863%), with       12 calls at 7.545833e+02 us per call
                 comms     = 0.620446 secs (  5.91%), with       12 calls at 5.170383e+04 us per call
     total accounted       = 10.495396 secs (   100%)
     total missing         = 0.000070 secs (0.000667%)
```

**develop**
```
   computeGaugeForceQuda Total time = 7.96804 secs
              download     = 4.945286 secs (  62.1%), with       24 calls at 2.060536e+05 us per call
                upload     = 0.827408 secs (  10.4%), with       12 calls at 6.895067e+04 us per call
                  init     = 0.014874 secs ( 0.187%), with       36 calls at 4.131667e+02 us per call
               compute     = 1.545521 secs (  19.4%), with       12 calls at 1.287934e+05 us per call
                  free     = 0.009129 secs ( 0.115%), with       12 calls at 7.607500e+02 us per call
                 comms     = 0.619028 secs (  7.77%), with       12 calls at 5.158567e+04 us per call
              constant     = 0.006710 secs (0.0842%), with       24 calls at 2.795833e+02 us per call
     total accounted       = 7.967956 secs (   100%)
     total missing         = 0.000086 secs (0.00108%)
```

**Comment by mikeaclark** [comment]
Thanks.  So the computation has gone down, it's just the upload that's got a lot slower.  I won't dwell on this, as it's likely an aberration.    This goes away entirely anyway when I get the resident momentum option enabled for mainline MILC.

For the dslash policy, I'll revert the change and keep it using the default policy.  I think I won't expose it either: I think what is best to do is to have each Dirac operator autotune the best policy to use.  I'll have to think about the best way to do that: this would ensure best performance regardless what machine you're on then.  Some work on the `GPU_COMMS` variant is needed also to let you choose at runtime whether to use or route though CPU memory, for full dynamic policy choice.  Anyway, work for another day.

**Comment by mathiaswagner** [comment]
Missed that previously: My run is 2+1 flavor.

**Pull request merged**